dotnet/src/Agents/Abstractions/AgentChat.cs

```diff
@@ -0,0 +1,184 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.SemanticKernel.Agents.Extensions;
+using Microsoft.SemanticKernel.Agents.Internal;
+using Microsoft.SemanticKernel.ChatCompletion;
+
+namespace Microsoft.SemanticKernel.Agents;
+
+/// <summary>
+/// Point of interaction for one or more agents.
+/// </summary>
+public abstract class AgentChat
+{
+    private readonly BroadcastQueue _broadcastQueue;
+    private readonly Dictionary<string, AgentChannel> _agentChannels;
+    private readonly Dictionary<Agent, string> _channelMap;
+    private readonly ChatHistory _history;
+
+    private int _isActive;
+
+    /// <summary>
+    /// Retrieve the message history, either the primary history or
+    /// an agent specific version.
+    /// </summary>
+    /// <param name="agent">An optional agent, if requesting an agent history.</param>
+    /// <param name="cancellationToken">The <see cref="CancellationToken"/> to monitor for cancellation requests. The default is <see cref="CancellationToken.None"/>.</param>
+    /// <returns>The message history</returns>
+    public IAsyncEnumerable<ChatMessageContent> GetHistoryAsync(Agent? agent = null, CancellationToken cancellationToken = default)
+    {
+        if (agent == null)
+        {
+            return this._history.ToDescendingAsync();
+        }
+
+        var channelKey = this.GetAgentHash(agent);
+        if (!this._agentChannels.TryGetValue(channelKey, out var channel))
+        {
+            return Array.Empty<ChatMessageContent>().ToAsyncEnumerable();
+        }
+
+        return channel.GetHistoryAsync(cancellationToken);
+    }
+
+    /// <summary>
+    /// Append messages to the conversation.
+    /// </summary>
+    /// <param name="message">Set of non-system messages with which to seed the conversation.</param>
+    public void AppendHistory(ChatMessageContent message)
+    {
+        this.AppendHistory(new[] { message });
+    }
+
+    /// <summary>
+    /// Append messages to the conversation.
+    /// </summary>
+    /// <param name="messages">Set of non-system messages with which to seed the conversation.</param>
+    /// <remarks>
+    /// Will throw KernelException if a system message is present, without taking any other action.
+    /// </remarks>
+    public void AppendHistory(IEnumerable<ChatMessageContent> messages)
+    {
+        bool hasSystemMessage = false;
+        var cleanMessages =
+            messages.Where(
+                m =>
+                {
+                    bool isSystemMessage = m.Role == AuthorRole.System;
+                    hasSystemMessage |= isSystemMessage;
+                    return !isSystemMessage;
+                }).ToArray();
```

Comment:
This creates another copy of the messages, is this needed? 
Can we instead check for the existence of non system messages and if none are found proceed with the original collection?